truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedTruffleRuntime.java

```diff
@@ -912,10 +913,45 @@ private void notifyCompilationFailure(OptimizedCallTarget callTarget, Throwable
     protected void onEngineCreated(EngineData engine) {
     }
 
+    private long stoppedCompilationTime = 0;
+    private boolean logShutdownCompilations = true;
+
     @SuppressWarnings("try")
     public CompilationTask submitForCompilation(OptimizedCallTarget optimizedCallTarget, boolean lastTierCompilation) {
-        Priority priority = new Priority(optimizedCallTarget.getCallAndLoopCount(), lastTierCompilation ? Priority.Tier.LAST : Priority.Tier.FIRST);
-        return getCompileQueue().submitCompilation(priority, optimizedCallTarget);
+        BackgroundCompileQueue queue = getCompileQueue();
+        CompilationActivityMode compilationActivityMode = getCompilationActivityMode();
+        if (compilationActivityMode == CompilationActivityMode.RUN_COMPILATION ||
+                (stoppedCompilationTime != 0 && System.currentTimeMillis() - stoppedCompilationTime > optimizedCallTarget.getOptionValue(StoppedCompilationRetryDelay))) {
+            stoppedCompilationTime = 0;
+            Priority priority = new Priority(optimizedCallTarget.getCallAndLoopCount(), lastTierCompilation ? Priority.Tier.LAST : Priority.Tier.FIRST);
+            return queue.submitCompilation(priority, optimizedCallTarget);
+        } else if (compilationActivityMode == CompilationActivityMode.STOP_COMPILATION) {
+            if (stoppedCompilationTime == 0) {
+                stoppedCompilationTime = System.currentTimeMillis();
+            }
+            // Flush the compilations queue. There's still a chance that compilation will be re-enabled
+            // eventually, if the hosts code cache can be cleaned up.
+            for (OptimizedCallTarget target : queue.getQueuedTargets(optimizedCallTarget.engine)) {
+                target.cancelCompilation("Compilation temporary disabled due to full code cache.");
+            }
+        } else {
+            // Compilation was shut down permanently because the hosts code cache ran full and
+            // the host was configured without support for code cache sweeping.
+            assert compilationActivityMode == CompilationActivityMode.SHUTDOWN_COMPILATION;
+            TruffleLogger logger = optimizedCallTarget.engine.getLogger("engine");
+            // The logger can be null if the engine is closed.
+            if (logger != null && logShutdownCompilations) {
+                logShutdownCompilations = false;
+                logger.log(Level.WARNING, "Truffle host compilations permanently disabled because of full code cache. "  +
```

Comment:
We do not really call Truffle compilations host compilations, that is typically reserved for compilations of HotSpot directly. Maybe we should just change it to Truffle compilations to avoid confusion.
Sidenote: it would not be a problem to show this warning once per engine.